src/libcore/pin.rs

```diff
@@ -403,13 +403,14 @@ pub struct Pin<P> {
     pointer: P,
 }
 
+// The following implementations aren't derived in order to avoid soundness
+// issues. `&self.pointer` should not be accessible to potentially malicious
+// trait implementations.
+//
+// See <https://internals.rust-lang.org/t/unsoundness-in-pin/11311/73> for more details.
+
 #[stable(feature = "pin_trait_impls", since = "1.41.0")]
-impl<P, Q> PartialEq<Pin<Q>> for Pin<P>
-where
-    P: Deref,
-    P::Target: PartialEq<Q::Target>,
-    Q: Deref,
-{
+impl<P: Deref<Target: PartialEq<Q::Target>>, Q: Deref> PartialEq<Pin<Q>> for Pin<P> {
```

Comment:
This isn't very readable and it is asymmetric, I preferred the more explicit style you used before... @Centril seems to disagree with me though.

Reply:
Well... I think this went overboard in this case. =P
You can still use ATB in `where` when it gets longer and keep it inline when it's shorter.

Reply:
Like this:
```rust
impl<P, Q> PartialEq<Pin<Q>> for Pin<P>
where
    Q: Deref,
    P: Deref<Target: PartialEq<Q::Target>>,
{
```
but in the case of `Hash` it's short enough to keep inline.

Balance™

Reply:
To be clear, my proposal is
```rust
impl<P: Deref, Q: Deref> PartialEq<Pin<Q>> for Pin<P>
  where P::Target: PartialEq<Q::Target>
```

Reply:
rustfmt dislikes `where P::Target` :)

Reply:
Your version is asymmetric though, and IMO it is bad style to express a symmetric fact in an asymmetric way.

What's rustfmt's problem with `P::Target`? I don't think we should let rustfmt bugs prevent us from improving readability.^^

Reply:
🤷‍♂ -- I think we're spending too much time on this. =P

> What's rustfmt's problem with `P::Target`? I don't think we should let rustfmt bugs prevent us from improving readability.^^

Well it will format it into the form I wrote above because that's what the style guide says it should look like... not my decision. ;)

Reply:
> Well it will format it into the form I wrote above because that's what the style guide says it should look like... not my decision. ;)

I [cannot reproduce that on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=893e20ffd7ae8fb292d26b2f43e4bb95)

> I think we're spending too much time on this. =P

You are probably right. ;)

Reply:
> I [cannot reproduce that on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=893e20ffd7ae8fb292d26b2f43e4bb95)

Oh, I was being literal about `where P` (same line!). :D

Reply:
I care less about the lines and more about making `P` and `Q`'s `Deref` requirement symmetric.

Reply:
@Centril so can we agree on this?
```
impl<P: Deref, Q: Deref> PartialEq<Pin<Q>> for Pin<P>
where
    P::Target: PartialEq<Q::Target>,
```

Reply:
@RalfJung I mainly think ATB is nice and clean, but if you feel strongly about not using it here then go for it because I don't feel strongly. :)

Reply:
I already changed this into

```rust
impl<P: Deref, Q: Deref> PartialEq<Pin<Q>> for Pin<P>
where
    P::Target: PartialEq<Q::Target>,
```

if that helps.